xdsl/irdl/declarative_assembly_format_parser.py

```diff
@@ -124,30 +127,34 @@ def parse_format(self) -> FormatProgram:
         elements: list[FormatDirective] = []
         while self._current_token.kind != Token.Kind.EOF:
             elements.append(self.parse_directive())
-            last = elements[-1]
-            if (
-                len(elements) >= 2
-                and isinstance(last, PunctuationDirective)
-                and last.punctuation == ","
-                and isinstance(
-                    elements[-2],
-                    VariadicOperandTypeDirective
-                    | VariadicResultTypeDirective
-                    | VariadicOperandVariable,
-                )
-            ):
-                self.raise_error(
-                    "A variadic directive cannot be followed by a comma literal."
-                )
 
         self.add_reserved_attrs_to_directive(elements)
         seen_variables = self.resolve_types()
+        self.verify_directives(elements)
         self.verify_attr_dict()
         self.verify_properties()
         self.verify_operands(seen_variables)
         self.verify_results(seen_variables)
         return FormatProgram(elements)
 
+    def verify_directives(self, elements: list[FormatDirective]):
+        for a, b in pairwise(elements):
+            match a, b:
+                case VariadicLikeFormatDirective(), PunctuationDirective(","):
+                    self.raise_error(
+                        "A variadic directive cannot be followed by a comma literal."
+                    )
+                case VariadicLikeTypeDirective(), VariadicLikeTypeDirective():
+                    self.raise_error(
+                        "A variadic type directive cannot be followed by another variadic type directive."
+                    )
```

Comment:
Sorry for bringing this up, but what happens if we have a `VariadicLikeTypeDirective` followed by a `VariadicOperandVariable` followed by another `VariadicLikeTypeDirective`? ;)
Or actually a `VariadicLikeFormatDirective` followed by an `Optional`, followed by a `PunctuationDirective(",")`?

If you want to implement the more complex algorithm, feel free, but otherwise let's just write a comment here to mark that this isn't supported, and maybe an issue!

Reply:
Ooooh :smiling_face_with_tear: It *did* feel too easy :sob: 
Given it's new features all around, what would you say? Do we merge things to start using it and make it bulletproof later, or do you want to make sure of this right now?
I don't have bandwidth for the bullet-proof option before a little while, I was just empty my procrastination output buffer ahah. If you have it and want it, feel very free to hijack this PR!

Reply:
Note to myself:
- VariadicOperand, VariadicsSmth/Comma, VariadicOperand is such an example

Reply:
Renote: Actually, MLIR doesn't even bother; let's just have some support for declarative syntax first and think about making it more robust once it at least works when it does! :slightly_smiling_face: